### PR TITLE
remove hard requirement that prevents more than 3 masters

### LINKED ISSUE #304: max master nodes
Gravity currently enforces a maximum on master nodes (3), without an apparent way to override.

This creates some implications in a couple cases:
- running 5 etcd nodes is a common operating method, to allow 2 nodes to fail before losing quorum
- When doing node replacements (such as AMI replacement), in some cases it may be preferable to add a node before deleting the node to be replaced. This doesn't really increase the redundancy, because at 4 nodes the cluster can only tolerate one failure, but if the cluster fails to expand it's not left stuck with 2/3 master nodes.

I think gravity should not rewrite master/node membership if explicitly defined in the application manifest, as this may also lead to unexpected behaviour.

Reference: https://github.com/gravitational/gravity/blob/master/lib/ops/opsservice/install.go#L885-L891

### PR DESCRIPTION
**Pull request by knisbet**
Fixes #304 

This change removes that hard requirement that prevents an administrator from adding more than 3 masters, even if specified in the manifest.

The reasons to allow more than 3 nodes:
- 5 nodes (n+2) is a common model for highly reliable systems
- When rotating nodes, it may be preferable to add the replacement node before deleting the node that is to be replaced. 

Note: I don't know why this was implemented as a hard requirement in the first place

### CODE REVIEW

File: lib/ops/opsservice/install.go [hunk 1]
```diff
@@ -911,22 +911,36 @@ func checkLicenseCPU(p licenseapi.Payload, numCPU uint) error {
...
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if profile.ServiceRole == schema.ServiceRoleMaster {
+			masters++
+		}
+	}
+
+	// assign the servers to their rolls
```
Comment (a-palchikov) [inline lib/ops/opsservice/install.go]: roles

**Comment by knisbet** [comment]
etcd performance does degrade with more members, this is correct. What happens is inserts to the raft log can only happen on the cluster leader, and the leader replicates to all members. So the more members the more time spent replicating additional log entries, which does have a performance impact.

However, gravity is in this weird situation, where if a user tries to specify more etcd members, for whatever reason, gravity just silently rewrites the node to not be a master. So even in a 7 node (n+3 redundancy) cluster, if a user wants to use a replacement strategy where they add then delete a node, we'll have some artificial limit that prevents that behaviour. Although I suppose running a 7 node cluster is far less common than a 3 or 5 node cluster.  

This is implemented as your soft limits approach, where if the node profile doesn't specify master or worker, the first 3 nodes in the cluster will become masters. It only doesn't create a hard cap on the masters if specified. If anything, having something other than 1/3/5/7 nodes should just be a warning in satellite.

**Comment by a-palchikov** [comment]
I think in this particular case, we should not have a hard limit on the number of masters. Satellite and documentation hints should suffice.
Honestly I'd be somewhat hesitant to place infrastructural limits that cannot be removed as opposed to scaling down the cluster if starts performing poorly due to initial misallocation of nodes.

**Comment by r0mant** [comment]
@a-palchikov @knisbet If you guys both think so, I will concede. I agree that the limit (7, 10, whatever) is somewhat artificial, but I also think that just removing the limit altogether without having an idea of how the system will scale/behave is asking for a trouble. At least we should try to make a large master-only cluster and see what happens (I don't think anyone tried that?).

**Comment by knisbet** [comment]
@r0mant I went to 9 masters using the telekube app, and I don't see anything weird, although the telekube app isn't very hard on the cluster.

# Single Node
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.010239s
PASS: Stddev is 0.000451s
PASS





# 2 nodes
knisbet-test1:/$ etcdctl3 member list
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 150 writes/s
PASS: Slowest request took 0.056850s
PASS: Stddev is 0.001135s
PASS


# 3 nodes
knisbet-test1:/$ etcdctl3 member list
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8b9ea8c629340b60, started, 10_162_15_249.competentstallman8178, https://10.162.15.249:2380, https://10.162.15.249:2379,https://10.162.15.249:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.040429s
PASS: Stddev is 0.001079s
PASS



# 4 nodes
knisbet-test1:/$ etcdctl3 member list
775ddc055485c103, started, 10_162_15_248.competentstallman8178, https://10.162.15.248:2380, https://10.162.15.248:2379,https://10.162.15.248:4001
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8b9ea8c629340b60, started, 10_162_15_249.competentstallman8178, https://10.162.15.249:2380, https://10.162.15.249:2379,https://10.162.15.249:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.015874s
PASS: Stddev is 0.000943s
PASS


# 5 nodes
knisbet-test1:/$ etcdctl3 member list
dab21782b63e68b, started, 10_162_15_245.competentstallman8178, https://10.162.15.245:2380, https://10.162.15.245:2379,https://10.162.15.245:4001
775ddc055485c103, started, 10_162_15_248.competentstallman8178, https://10.162.15.248:2380, https://10.162.15.248:2379,https://10.162.15.248:4001
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8b9ea8c629340b60, started, 10_162_15_249.competentstallman8178, https://10.162.15.249:2380, https://10.162.15.249:2379,https://10.162.15.249:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.018748s
PASS: Stddev is 0.000895s
PASS

# 6 nodes
knisbet-test1:/$ etcdctl3 member list
dab21782b63e68b, started, 10_162_15_245.competentstallman8178, https://10.162.15.245:2380, https://10.162.15.245:2379,https://10.162.15.245:4001
775ddc055485c103, started, 10_162_15_248.competentstallman8178, https://10.162.15.248:2380, https://10.162.15.248:2379,https://10.162.15.248:4001
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8b9ea8c629340b60, started, 10_162_15_249.competentstallman8178, https://10.162.15.249:2380, https://10.162.15.249:2379,https://10.162.15.249:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
f3229338dd572a7c, started, 10_162_15_246.competentstallman8178, https://10.162.15.246:2380, https://10.162.15.246:2379,https://10.162.15.246:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.021954s
PASS: Stddev is 0.000811s
PASS


# 7 nodes
knisbet-test1:/$ etcdctl3 member list
dab21782b63e68b, started, 10_162_15_245.competentstallman8178, https://10.162.15.245:2380, https://10.162.15.245:2379,https://10.162.15.245:4001
775ddc055485c103, started, 10_162_15_248.competentstallman8178, https://10.162.15.248:2380, https://10.162.15.248:2379,https://10.162.15.248:4001
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8b9ea8c629340b60, started, 10_162_15_249.competentstallman8178, https://10.162.15.249:2380, https://10.162.15.249:2379,https://10.162.15.249:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
dd27a3bca3ec7ba5, started, 10_162_15_247.competentstallman8178, https://10.162.15.247:2380, https://10.162.15.247:2379,https://10.162.15.247:4001
f3229338dd572a7c, started, 10_162_15_246.competentstallman8178, https://10.162.15.246:2380, https://10.162.15.246:2379,https://10.162.15.246:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.017133s
PASS: Stddev is 0.000747s
PASS
knisbet-test1:/$



# 8 nodes
knisbet-test1:/$ etcdctl3 member list
dab21782b63e68b, started, 10_162_15_245.competentstallman8178, https://10.162.15.245:2380, https://10.162.15.245:2379,https://10.162.15.245:4001
16570b444b40996c, started, 10_162_15_228.competentstallman8178, https://10.162.15.228:2380, https://10.162.15.228:2379,https://10.162.15.228:4001
775ddc055485c103, started, 10_162_15_248.competentstallman8178, https://10.162.15.248:2380, https://10.162.15.248:2379,https://10.162.15.248:4001
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8b9ea8c629340b60, started, 10_162_15_249.competentstallman8178, https://10.162.15.249:2380, https://10.162.15.249:2379,https://10.162.15.249:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
dd27a3bca3ec7ba5, started, 10_162_15_247.competentstallman8178, https://10.162.15.247:2380, https://10.162.15.247:2379,https://10.162.15.247:4001
f3229338dd572a7c, started, 10_162_15_246.competentstallman8178, https://10.162.15.246:2380, https://10.162.15.246:2379,https://10.162.15.246:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 151 writes/s
PASS: Slowest request took 0.015699s
PASS: Stddev is 0.000704s
PASS



# 9 nodes
knisbet-test1:/$ etcdctl3 member list
dab21782b63e68b, started, 10_162_15_245.competentstallman8178, https://10.162.15.245:2380, https://10.162.15.245:2379,https://10.162.15.245:4001
13d758505fa4f1d1, started, 10_162_15_242.competentstallman8178, https://10.162.15.242:2380, https://10.162.15.242:2379,https://10.162.15.242:4001
16570b444b40996c, started, 10_162_15_228.competentstallman8178, https://10.162.15.228:2380, https://10.162.15.228:2379,https://10.162.15.228:4001
775ddc055485c103, started, 10_162_15_248.competentstallman8178, https://10.162.15.248:2380, https://10.162.15.248:2379,https://10.162.15.248:4001
823b2103c2bb5159, started, 10_162_15_243.competentstallman8178, https://10.162.15.243:2380, https://10.162.15.243:2379,https://10.162.15.243:4001
8b9ea8c629340b60, started, 10_162_15_249.competentstallman8178, https://10.162.15.249:2380, https://10.162.15.249:2379,https://10.162.15.249:4001
8c83715ef7b1264b, started, 10_162_15_244.competentstallman8178, https://10.162.15.244:2380, https://10.162.15.244:2379,https://10.162.15.244:4001
dd27a3bca3ec7ba5, started, 10_162_15_247.competentstallman8178, https://10.162.15.247:2380, https://10.162.15.247:2379,https://10.162.15.247:4001
f3229338dd572a7c, started, 10_162_15_246.competentstallman8178, https://10.162.15.246:2380, https://10.162.15.246:2379,https://10.162.15.246:4001
knisbet-test1:/$ etcdctl3 check perf
 60 / 60 Booooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooo! 100.00%1m0s
PASS: Throughput is 150 writes/s
PASS: Slowest request took 0.014148s
PASS: Stddev is 0.000634s
PASS

**Pull request merged**